Fix BlockEvent.PlaceEvent ignoring EnumHand (#3221)

Choonster opened the pull request:
This fixes #3221.

Choonster committed:
Fix BlockEvent.PlaceEvent ignoring EnumHand (#3221)

Actuarius commented:
@mezz added labels [Bug]

jredfox commented:
@Choonster 
I noticed in this code that You used the main hand and blockstate 0 for multiblock place. Are you sure you couldn't have offhand as well or would that be more work? I also noticed in previous files that blockstate had an if statement for blockstate 1 did you miss this for your mutliblock place method?

 this(blockSnapshots, placedAgainst, player, EnumHand.MAIN_HAND);
-        }
  +
-        public MultiPlaceEvent(List<BlockSnapshot> blockSnapshots, IBlockState placedAgainst, EntityPlayer player, EnumHand hand) {
-            super(blockSnapshots.get(0), placedAgainst, player, hand);

Choonster commented:
I only use the main hand for the deprecated overloads of the event constructors and factory methods to preserve the old behaviour for mods that use them. The new overloads used by Forge take an `EnumHand` argument.

The new code uses `blocksSnapshots.get(0)` in exactly the same way as the old code: to get the snapshot when there's only one snapshot, to get the `placedAgainst` state when there's multiple snapshots and to pass the first snapshot to the `PlaceEvent` constructor from the `MultiPlaceEvent` constructor.

Just like before, mods that subscribe to `MultiPlaceEvent` will have access to all the snapshots (if there was more than one) and mods that subscribe to `PlaceEvent` will have access to the first snapshot.

jredfox commented:
When Will this be implemented? I would like to get a working version of my mod done. It seems Like they checked it says no conflicts and it's been like a week.

jredfox commented:
Don't mean to be rude but, if all checks have passed no conflicts with it why hasn't it been implemented? I patentially waited a week and saw a forge update got excited then realized that this fix wasn't in there.  **This is critical with doing any advanced mod coding.  So could you put this in the next version please and make this a priority**?  @LexManos

gr8pefish commented:
I too would like to see this implemented. @Choonster has provided a good, clean implementation of a necessary addendum to the event.

jredfox commented:
Well then it's agreed then that this is good clean and bug free then and the code is provided then forge now needs to add this in the next build. Mods won't be the same without this fix.

LexManos commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -112,6 +112,10 @@
+     */
+    @Deprecated
     public static MultiPlaceEvent onPlayerMultiBlockPlace(EntityPlayer player, List<BlockSnapshot> blockSnapshots, EnumFacing direction)
     {
```
Dont duplicate code, bounce to the new one with the argument null, this makes it more obvious what to remove when deleting deprecated functions. And illustrates the parameter CAN be null.

LexManos commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -129,6 +146,14 @@ public static PlaceEvent onPlayerBlockPlace(EntityPlayer player, BlockSnapshot b
     }
 
+    public static PlaceEvent onPlayerBlockPlace(EntityPlayer player, BlockSnapshot blockSnapshot, EnumFacing direction, EnumHand hand)
+    {
```
Same as above

LexManos commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -170,18 +171,25 @@ public void setExpToDrop(int exp)
             super(blockSnapshot.getWorld(), blockSnapshot.getPos(), blockSnapshot.getCurrentBlock());
             this.player = player;
-            this.itemInHand = player.getHeldItemMainhand();
+            this.itemInHand = player.getHeldItem(hand);
```
You may need null protection here.

LexManos commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -190,6 +198,7 @@ public PlaceEvent(BlockSnapshot blockSnapshot, IBlockState placedAgainst, Entity
         public BlockSnapshot getBlockSnapshot() { return blockSnapshot; }
         public IBlockState getPlacedBlock() { return placedBlock; }
         public IBlockState getPlacedAgainst() { return placedAgainst; }
+        public EnumHand getHand() { return hand; }
```
@Nullable

Choonster committed:
Clean up BlockEvent.PlaceEvent EnumHand changes

- Replace bodies of decprecated factory methods with calls to the new overloads
- Don't call EntityLivingBase#getHeldItem with a null EnumHand
- Add @Nullable annotations

Choonster commented:
The new commit should address your comments.

gr8pefish commented:
If anything else is needed, I am willing to help out in whatever regard I can to make this PR be pulled :)

gr8pefish commented:
Any word on if this will be accepted? I just looked over the code myself and noted no issues with it; seeing as it addresses an issue I would like to see it merged in the near future if possible. Or at least an explanation of what is causing a holdup (if anything).

That being said, if it isn't ever going to be merged, I would simply like to know that so I can start writing my own workarounds. Thanks!
